main.cpp

```diff
@@ -68,29 +90,45 @@ int main(int argc, char *argv[])
 #endif
     run<double>();
 
+  // End MPI
+#ifdef USE_MPI
+  MPI_Finalize();
+#endif
 }
 
 template <typename T>
 void run()
 {
+#ifdef USE_MPI
+  if (rank == 0)
+#endif
+
+  {
   std::cout << "Running kernels " << num_times << " times" << std::endl;
 
   if (sizeof(T) == sizeof(float))
     std::cout << "Precision: float" << std::endl;
   else
     std::cout << "Precision: double" << std::endl;
+  }
 
   // Create host vectors
   std::vector<T> a(ARRAY_SIZE);
   std::vector<T> b(ARRAY_SIZE);
   std::vector<T> c(ARRAY_SIZE);
+
+#ifdef USE_MPI
+if (rank == 0)
+#endif
+  {
   std::streamsize ss = std::cout.precision();
   std::cout << std::setprecision(1) << std::fixed
     << "Array size: " << ARRAY_SIZE*sizeof(T)*1.0E-6 << " MB"
     << " (=" << ARRAY_SIZE*sizeof(T)*1.0E-9 << " GB)" << std::endl;
   std::cout << "Total size: " << 3.0*ARRAY_SIZE*sizeof(T)*1.0E-6 << " MB"
     << " (=" << 3.0*ARRAY_SIZE*sizeof(T)*1.0E-9 << " GB)" << std::endl;
   std::cout.precision(ss);
+  }
```

Comment:
Not really, I think this is about as good at can get:

    std::cout << std::setprecision(1) << std::fixed
    #ifdef USE_MPI
      << "Array size (per MPI rank): "
    #else
      << "Array size: "
    #endif
      << "...";